File: backend/api/models/expense.py
from api.database import db
from sqlalchemy import (
    Column, Integer, String, REAL, Enum, TIMESTAMP, 
    ForeignKeyConstraint, PrimaryKeyConstraint 
)
import enum
from datetime import datetime

class SplitType(enum.Enum):
    EQUAL = "equal"
    EXACT = "exact"
    PERCENTAGE = "percentage"

class Expense(db.Model):
    __tablename__ = 'expense'
    
    expense_id = Column(Integer, primary_key=True, autoincrement=True)
    title = Column(String(50))
    total = Column(REAL, nullable=False)
    payer = Column(Integer, nullable=False)
    group_id = Column(Integer, nullable=False)
    split = Column(Enum(SplitType), nullable=False)
    timestamp = Column(TIMESTAMP(timezone=True), nullable=False, default=datetime.utcnow)

    __table_args__ = (
        ForeignKeyConstraint(['group_id'], ['group.group_id']),
        ForeignKeyConstraint(['payer'], ['user.user_id'])
    )

class ExpenseShare(db.Model):
    __tablename__ = 'expense_share'

    expense_id = Column(Integer)
    user_id = Column(Integer)
    credit = Column(REAL, nullable=False)
    debit = Column(REAL, nullable=False)
    share = Column(REAL)

    __table_args__ = (
        PrimaryKeyConstraint('expense_id', 'user_id'),
        ForeignKeyConstraint(['user_id'], ['user.user_id'])
    )

File: frontend/src/app/models/expense.ts
import { propertyMap } from '../_helpers/model-mapper';

export enum SplitType {
    EQUAL = 'equal',
    EXACT = 'exact',
    PERCENTAGE = 'percentage'
}

export class Expense {
    @propertyMap('expense_id')
    id: number;
    title: string;
    total: number;
    payer: number;
    @propertyMap('group_id')
    groupId: number;
    split: SplitType;
    timestamp: string;
    // TODO: user userid instead
    shareMap: Map<any, any>;

    constructor(id?: number, groupId?: number) {
        this.id = id;
        this.groupId = groupId;
        this.shareMap = new Map();
    }

    // TODO: convert to generic util function
    toCustomObj() {
        return {
            id: this.id,
            title: this.title,
            total: this.total,
            payer: this.payer,
            groupId: this.groupId,
            split: this.split,
            shareMap: this.mapToObject(this.shareMap)
        };
    }

    private mapToObject(inputMap) {
        const obj = {};
        inputMap.forEach((value, key) => {
            obj[key] = value;
        });
        return obj;
    }
}
